### PR TITLE
Add PCL as a target language for pulumi convert

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

Every time I need to debug `pulumi convert` I have to hack YAML program-gen and make it output the intermediate PCL program text to stdout and work from there. Here, I've added PCL as a proper target language for `pulumi convert` (without needing to parse the program text and generate `pcl.Program` from it) 

I tried a simpler approach:
```go
projectGenerator = func(directory string, project workspace.Project, p *pcl.Program) error {
	programText := fmt.Sprintf("%v", p)
	outputFile := path.Join(directory, "main.pp")
	err := ioutil.WriteFile(outputFile, []byte(programText), 0600)
	if err != nil {
		return fmt.Errorf("could not write output program: %w", err)
	}
	return nil
}
```
but the default formatter for `pcl.Program` writes out invalid PCL (including non-formatted type metadata) so I went with this approach instead 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/cmd/pulumi/convert.go [hunk 1]
```diff
@@ -94,6 +134,10 @@ func runConvert(cwd string, language string, outDir string, generateOnly bool) r
 		projectGenerator = javagen.GenerateProject
 	case "yaml": // nolint: goconst
 		projectGenerator = yamlgen.GenerateProject
+	case "pulumi", "pcl":
```
Comment (iwahbe) [inline pkg/cmd/pulumi/convert.go:137]: I would consider adding `fmt.Fprintf(os.Stderr, "Warning: PCL is an implementation feature and not stable\n")` or gating behind a env var to make sure people know this is not for normal use.

**Review by iwahbe — APPROVED** [review]
This will definitely be helpful. I wouldn't add a changelog entry, since this is not really external facing.

File: pkg/cmd/pulumi/convert.go (hunk 1 above)
Comment (Frassle) [reply pkg/cmd/pulumi/convert.go:137]: PULUMI_DEV is used elsewhere to silence warnings like this

Comment (Zaid-Ajaj) [reply pkg/cmd/pulumi/convert.go:137]: I've updated the code such that this only runs when `PULUMI_DEV` is present and is truthy ✅

File: pkg/cmd/pulumi/convert.go [hunk 2]
```diff
@@ -135,9 +135,14 @@ func runConvert(cwd string, language string, outDir string, generateOnly bool) r
...
-		// since we don't need Eject to get the full program,
-		// we can just convert the YAML directly to PCL
-		return runConvertPcl(cwd, outDir)
+		if cmdutil.IsTruthy(os.Getenv("PULUMI_DEV")) {
+			// since we don't need Eject to get the full program,
+			// we can just convert the YAML directly to PCL
+			return runConvertPcl(cwd, outDir)
+		}
+
+		return result.Errorf("cannot generate programs for %q language", language)
```
Comment (iwahbe) [inline pkg/cmd/pulumi/convert.go:144]: I think this can just be `fallthrough`, but I'm never sure exactly when that's valid.

Comment (Zaid-Ajaj) [reply pkg/cmd/pulumi/convert.go]: First time I read about this, added and tested. Works like a charm ✅

**Comment by Zaid-Ajaj** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors cancel

**Comment by AaronFriel** [comment]
bors merge

**Comment by Zaid-Ajaj** [comment]
bors merge

**Pull request merged**